Implemented saving of new messages in channels bot is added to.

khalidsharmarke opened the pull request:
standardized client handling of creating requests to Django api.
modified ORM's relational mapping between tables to reflect intended query-able ability.

khalidsharmarke committed:
Implemented saving of new messages in channels bot is added to.
standardized client handling of creating requests to Django api.
modified ORM's relational mapping between tables to reflect intended query-able ability.

willwen commented on client/api_requests/IncomingDataHandlers/BaseHandler.js:
```diff
@@ -1,16 +1,20 @@
 		const item = constructor(data);
 		return {
-			[`${constructor.name.toLowerCase()}`]: item,
+			[`${constructor.name.toLowerCase()}s`]: [item],
```
add comment? whats the 's' for

willwen commented on client/api_requests/IncomingDataHandlers/ChannelHandler.js:
```diff
@@ -1,16 +1,11 @@
-import { Channel } from '../schema_definitions.js';
+import { Channel } from '../schemaDefinitions.js';
+
+const endpoint = '/channel';
```
usually consts are in all caps
so 
```
const ENDPOINT= '/channel';
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const#basic_const_usage

willwen commented on client/api_requests/IncomingDataHandlers/MessageHandler.js:
```diff
@@ -1,7 +1,12 @@
+import { Message } from '../schemaDefinitions.js';
 
-class MessageHandler {}
+const endpoint = '/message';
```
same here

willwen commented on client/api_requests/populateDB.js:
```diff
@@ -53,10 +53,20 @@ async function populateDB(app_method, constructor, endpoint) {
 	let promisesArr = [];
 	while (true) {
```
you can consolidate:
```
while cursor != '' :
    do stuff
    cursor = 
```

khalidsharmarke replied:
we actually cant do that unless we specify a do while loop.
first cursor is always ''.
we dont get first cursor untill first request so wont execute unless do...while loop.

willwen replied:
https://dev.to/andre347/how-to-use-a-do-while-loop-for-api-pagination-375b

willwen commented on slackbot-api/api/models/Message.py:
```diff
@@ -3,8 +3,8 @@
 class Message(models.Model):
-    user_id = models.ForeignKey(User, on_delete=models.PROTECT)
-    channel_id = models.ForeignKey(Channel, on_delete=models.PROTECT)
+    user = models.ForeignKey(User, to_field='user_id', on_delete=models.PROTECT)
```
i would keep it as  user_id and channel_id

khalidsharmarke replied:
its bad practice to do so.
The ORM will replace the fields that reference foreign_keys with fields from the other Models
it seems best to just name the reference itself as the field name

khalidsharmarke replied:
more explicitly, per instance, the ORM will need an id to be passed for referencing the table (the 'to_field' argument) when creating a new one. since we are looking to pass around only IDs for creating this behavior is fine.
When accessing the data of a given Message, it'd be misleading to say Message.user_id.foo_bar when your looking to actually access the User for said Message.

under the hood, the DB is still behaving the intended way of only having the actual 'id' (not the incremented pk) be saved for this column for both tables while still naming the column the 'to_field' argument passed which is what I consider also fine behavior.

willwen commented on slackbot-api/api/views/BaseView.py:
```diff
@@ -3,7 +3,8 @@
 
 from api.views.ErrorHandler import ErrorHandler
-from django.db.utils import IntegrityError
+from  api.models.User import User
```
spacing

willwen commented on slackbot-api/api/views/BaseView.py:
```diff
@@ -19,30 +20,12 @@ def edit_item(item_id, item_definition, item_params):
     def post_method(self, request, item_name, item_definition):
-        item_name_as_plural = item_name + 's'
         try:
+            item_name_as_plural = item_name + 's'
```
ehhhhhhprobably made it more confusing than needed

willwen reviewed (commented):
can you add what you did to test  and validate this new code works

khalidsharmarke commented:
Testing of new feature includes:
Dropping of current data in DB
running client-side to preload DB User and Channel tables
manually posting a message to a slack channel the bot is apart of
following the requests using a series of log statements on both client and api sides
manually checking DB for the saved message to correlate with intended behavior.

khalidsharmarke committed:
brought code to standard.
added comments to explain implementation details.
capitalized const variables to conform to common standard.
removed unused imports.
